src/common/game_manager.h

```diff
@@ -0,0 +1,51 @@
+#pragma once
+
+#include <functional>
+#include <thread>
+#include <core/provider.pb.h>
+
+#include "event_loop.h"
+#include "network_interface.h"
+#include "state_machine.h"
+
+namespace SailGame { namespace Common {
+
+using Core::ProviderMsg;
+
+template<typename StateT>
+class GameManager {
+public:
+    GameManager()
+        : mStateMachine(std::make_unique<StateMachine<StateT>>())
+    {
+        auto eventHappensCallback = [this](const ProviderMsgPtr &event) {
+            mEventLoop->AppendEvent(event);
+        };
+        mNetworkInterface = std::make_unique<NetworkInterface>(
+            "localhost:50051", eventHappensCallback);
+
+        auto eventProcessedCallback = [this](const ProviderMsgPtr &event) {
+            ProcessEvent(event);
+        };
+        mEventLoop = std::make_unique<EventLoop>(eventProcessedCallback);
```

Comment:
NetworkInterface and mEventLoop could be member variable,  don't understand why you need unique_pointer.

also,  the two callback can be member functions.

Reply:
Given the fact that GameManager, EventLoop, StateMachine and NetworkInterface are designed to be generic lib (all in `common` directory), GameManager holds pointer to them so that we can pass in **derivate classes** in the future.

Reply:
Also, if you mean something like
```c++
using namespace std::placeholders;
mEventLoop = std::make_unique<EventLoop>(std::bind(&GameManager::ProcessEvent, this, _1));
```
I think using lambda is more straight-forward and readable

Reply:
Try to use EventLoop& in that case.

Reply:
member function is more readable.
Fine, I don't like to argue which is better.
I would say both of them is shit.

the interface is terrible.   If you pass the callback by constructor.
how can you make Evenloop generic? and pass it by pointer?

Reply:
class EventLoopListener {
    virtual ProcessedEvent(const & event) = 0;
}

class EventLoop{
      AddListener(EventLoopListener *) { mSet.insert()}
      RemoveListener(EventLoopListener *) { mSet.remove()}
      
      process() {  for mSet { ProcessedEvent() }}
}

class GameManager : public EventLoopListener {
      GameManager(EventLoop& e) : mEventLoop(e) 
{
   mEventLoop.AddListener(this);
}

ProcessedEvent(const & event) override {}
}

Reply:
awesome